src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+                                        accessIDtoDataLocality_.find(oldID)->second);
+    }
+
+    // Create the cache-filler stage
+    std::vector<int> oldIDs;
+    std::vector<int> newIDs;
+    for(auto& idPair : oldIDtoNewID) {
+      bool hasReadBeforeWrite = checkReadBeforeWrite(idPair.second);
+      if(hasReadBeforeWrite) {
+        oldIDs.push_back(idPair.first);
+        newIDs.push_back(idPair.second);
+      }
+    }
+    if(oldIDs.size() > 0) {
+      auto stageBegin = multiStagePrt_->getStages().begin();
+
+      std::shared_ptr<Stage> cacheFillStage =
+          createAssignmentStage(multiStagePrt_->getEnclosingInterval(), newIDs, oldIDs);
+      multiStagePrt_->getStages().insert(stageBegin, std::move(cacheFillStage));
+    }
+  }
+
+  /// @brief we create one stage at the end of the multistage that flushes back all the variables
+  /// that we cached in temporaries
+  void addFlushStages() {
+    std::vector<int> oldIDs;
+    std::vector<int> newIDs;
+    for(auto& idPair : oldIDtoNewID) {
+      if(!checkReadOnlyAccess(idPair.second)) {
+        oldIDs.push_back(idPair.first);
+        newIDs.push_back(idPair.second);
+      }
+    }
+
+    if(oldIDs.size() > 0) {
+      auto stageEnd = multiStagePrt_->getStages().end();
+
+      std::shared_ptr<Stage> cacheFlushStage =
+          createAssignmentStage(multiStagePrt_->getEnclosingInterval(), oldIDs, newIDs);
+
+      // Insert the new stage at the found location
+      multiStagePrt_->getStages().insert(stageEnd, std::move(cacheFlushStage));
+    }
+  }
+
+  /// @brief creates the stage in which assignment happens (fill and flush)
+  std::shared_ptr<Stage> createAssignmentStage(const Interval& interval,
+                                               const std::vector<int>& assignmentIDs,
+                                               const std::vector<int>& assigneeIDs) {
+    // Add the cache Flush stage
+    std::shared_ptr<Stage> assignmentStage = std::make_shared<Stage>(
+        instantiation_, multiStagePrt_, instantiation_->nextUID(), interval);
+    std::unique_ptr<DoMethod> domethod = make_unique<DoMethod>(assignmentStage.get(), interval);
+    domethod->getStatementAccessesPairs().clear();
+
+    for(int i = 0; i < assignmentIDs.size(); ++i) {
+      int assignmentID = assignmentIDs[i];
+      int assigneeID = assigneeIDs[i];
+      addAssignmentToDoMethod(domethod, assignmentID, assigneeID);
+    }
+
+    // Add the single do method to the new Stage
+    assignmentStage->getDoMethods().clear();
+    assignmentStage->addDoMethod(domethod);
+    assignmentStage->update();
+
+    return assignmentStage;
+  }
+
+  ///@brief add the assignment operator of two unique id's to a given domethod
+  void addAssignmentToDoMethod(std::unique_ptr<DoMethod>& domethod, int assignmentID,
+                               int assigneeID) {
+    // Create the StatementAccessPair of the assignment with the new and old variables
+    std::shared_ptr<FieldAccessExpr> fa_assignee =
```

Comment:
It is very hard to follow these few lines...  It would strongly recommend using `auto` in this case as the syntactic noise from `shared_ptr` here is crazy.

Reply:
after changing it I am also not super pleased with how auto looks. In either case this is a noisy part of the code as we manually create a stage in memory. Whatever works better